client/landing/gutenboarding/index.tsx

```diff
@@ -139,11 +139,9 @@ function waitForSelectedSite(): Promise< Site | undefined > {
  */
 function ensureRandomizedDesignsAreUpToDate() {
 	const designsInStore = select( ONBOARD_STORE ).getRandomizedDesigns();
+	const availableDesigns = getAvailableDesigns();
 	if ( ! isDeepEqual( designsInStore.featured, availableDesigns.featured ) ) {
-		dispatch( ONBOARD_STORE ).setRandomizedDesigns( {
-			...availableDesigns,
-			featured: shuffle( availableDesigns.featured ),
-		} );
+		dispatch( ONBOARD_STORE ).setRandomizedDesigns( getAvailableDesigns( { randomize: true } ) );
```

Comment:
This PR only aims at refactoring the shuffling logic and moving it directly to the `@automattic/design-picker` package.

But as I've been working on the refactor, I'm not entirely sure about the logic of this function.

The `designsInStore` are supposed to be already randomized, while `availableDesigns` are not randomized.

Therefore, the check `if ( ! isDeepEqual( designsInStore.featured, availableDesigns.featured ) ) { ... }` will always return `true`, unless the persisted `designsInStore` are for some reason not actually randomized. Which means that `ensureRandomizedDesignsAreUpToDate()` doesn't _really_ ensure that the designs are up to date — the designs are randomized every time this  function is called.

Am i missing something?

Reply:
`isDeepEqual` is using `xorwith` as an equivalent to sorting the top layer before comparison. It will return true if every design in one has a matching design in the other and vice versa (using `_.isEqual` for deep comparison).

That means the overall check will return true if any designs are changed/added/removed which seems correct, right?

Reply:
Alternatively, my super-naive performance testing suggests that `isEqual( keyBy( designsInStore.featured, 'slug' ), keyBy( availableDesigns.featured, 'slug' ) )` is an order of magnitude faster that `isDeepEqual`:

```
const start = performance.now();
const comparison = isDeepEqual( designsInStore.featured, availableDesigns.featured );
const isDeepEqualFinished = performance.now();
const keyComparison = isEqual( keyBy( designsInStore.featured, 'slug' ), keyBy( availableDesigns.featured, 'slug' ) );
const keyComparisonFinished = performance.now();
console.log( 'comparison:', comparison, isDeepEqualFinished - start );
console.log( 'keyComparison:', keyComparison, keyComparisonFinished - isDeepEqualFinished );
```

![WordPress_com](https://user-images.githubusercontent.com/5952255/113792423-46425c80-9789-11eb-84b3-99076df4aa72.jpg)

Reply:
Good point! I took your suggestion and re-wrote it in a way that doesn't require `lodash` — it seems to work fine.

Also changed function name and some of the comments to clarify a bit better what's going on.

Relevant commit: https://github.com/Automattic/wp-calypso/pull/51633/commits/80a337d6657fe3037cbd591904bd802859e4a2eb